builder/xenserver/common/step_wait_for_ip.go

```diff
@@ -29,47 +29,62 @@ func (self *StepWaitForIP) Run(ctx context.Context, state multistep.StateBag) mu
 		return multistep.ActionHalt
 	}
 
-	var ip string
-	err = InterruptibleWait{
-		Timeout:           self.Timeout,
-		PredicateInterval: 5 * time.Second,
-		Predicate: func() (result bool, err error) {
+	go func(c Connection, ui packer.Ui, config CommonConfig) {
+		state.Put("instance_ssh_address", "")
+		var ip string
 
+		for true {
+			time.Sleep(500 * time.Millisecond)
 			if config.IPGetter == "auto" || config.IPGetter == "http" {
-
 				// Snoop IP from HTTP fetch
 				select {
-				case ip = <-self.Chan:
-					ui.Message(fmt.Sprintf("Got IP '%s' from HTTP request", ip))
-					return true, nil
+				case temp_ip := <-self.Chan:
+					if ip != temp_ip {
+						ip = temp_ip
+						ui.Message(fmt.Sprintf("Got IP '%s' from HTTP request", ip))
+						state.Put("instance_ssh_address", ip)
+					}
 				default:
 				}
-
 			}
 
 			if config.IPGetter == "auto" || config.IPGetter == "tools" {
-
 				// Look for PV IP
 				m, err := c.client.VM.GetGuestMetrics(c.session, instance)
 				if err != nil {
-					return false, err
+					continue
 				}
 				if m != "" {
 					metrics, err := c.client.VMGuestMetrics.GetRecord(c.session, m)
 					if err != nil {
-						return false, err
+						continue
 					}
 					networks := metrics.Networks
-					var ok bool
-					if ip, ok = networks["0/ip"]; ok {
-						if ip != "" {
+					if temp_ip, ok := networks["0/ip"]; ok {
+						if temp_ip != "" && ip != temp_ip {
+							ip = temp_ip
 							ui.Message(fmt.Sprintf("Got IP '%s' from XenServer tools", ip))
-							return true, nil
+							state.Put("instance_ssh_address", ip)
 						}
 					}
 				}
 
 			}
+		}
+	}(*c, ui, config)
+
+	time.Sleep(500 * time.Millisecond)
```

Comment:
What's the purpose of this `time.Sleep`?

Reply:
The purpose of this `time.sleep` is, as you mentioned in your other comment, a minimum wait time to prevent the `Predicate` from resolving early. I have replaced the hardcoded value with a configurable one instead.